pm/recipient.go

```diff
@@ -0,0 +1,223 @@
+// receives tickets and redeems winning tickets
+type recipient struct {
+	val    Validator
+	broker Broker
+	store  TicketStore
+
+	secret [32]byte
+
+	invalidRands sync.Map
+
+	senderNonces     map[string]uint64
+	senderNoncesLock sync.Mutex
+
+	faceValue *big.Int
+	winProb   *big.Int
+}
+
+// NewRecipient creates an instance of a recipient with an
+// automatically generated random secret
+func NewRecipient(broker Broker, val Validator, store TicketStore, faceValue *big.Int, winProb *big.Int) (Recipient, error) {
+	randBytes := make([]byte, 32)
+	if _, err := rand.Read(randBytes); err != nil {
+		return nil, err
+	}
+
+	var secret [32]byte
+	copy(secret[:], randBytes[:32])
+
+	return NewRecipientWithSecret(broker, val, store, secret, faceValue, winProb), nil
+}
+
+// NewRecipientWithSecret creates an instance of a recipient with a user provided
+// secret. In most cases, NewRecipient should be used instead which will
+// automatically generate a random secret
+func NewRecipientWithSecret(broker Broker, val Validator, store TicketStore, secret [32]byte, faceValue *big.Int, winProb *big.Int) Recipient {
+	return &recipient{
+		broker:       broker,
+		val:          val,
+		store:        store,
+		secret:       secret,
+		faceValue:    faceValue,
+		senderNonces: make(map[string]uint64),
+		winProb:      winProb,
+	}
+}
+
+// ReceiveTicket validates and processes a received ticket
+func (r *recipient) ReceiveTicket(ticket *Ticket, sig []byte, seed *big.Int) (bool, error) {
+	recipientRand := r.rand(seed, ticket.Sender)
+
+	if crypto.Keccak256Hash(ethcommon.LeftPadBytes(recipientRand.Bytes(), uint256Size)) != ticket.RecipientRandHash {
+		return false, errors.Errorf("invalid recipientRand generated from seed %v", seed)
+	}
+
+	if ticket.FaceValue.Cmp(r.faceValue) != 0 {
+		return false, errors.Errorf("invalid ticket faceValue %v", ticket.FaceValue)
+	}
+
+	if ticket.WinProb.Cmp(r.winProb) != 0 {
+		return false, errors.Errorf("invalid ticket winProb %v", ticket.WinProb)
+	}
+
+	if err := r.val.ValidateTicket(ticket, sig, recipientRand); err != nil {
+		return false, err
+	}
+
+	if !r.validRand(recipientRand) {
+		return false, errors.Errorf("invalid already revealed recipientRand %v", recipientRand)
+	}
+
+	if err := r.updateSenderNonce(recipientRand, ticket.SenderNonce); err != nil {
+		return false, err
+	}
+
+	if r.val.IsWinningTicket(ticket, sig, recipientRand) {
+		if err := r.store.Store(ticket.RecipientRandHash.Hex(), ticket, sig, recipientRand); err != nil {
+			return true, err
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// RedeemWinningTicket redeems all winning tickets with the broker
+// for a session ID
+func (r *recipient) RedeemWinningTickets(sessionID string) error {
+	tickets, sigs, recipientRands, err := r.store.Load(sessionID)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(tickets); i++ {
+		if err := r.redeemWinningTicket(tickets[i], sigs[i], recipientRands[i]); err != nil {
+			return err
```

Comment:
Wondering - perhaps it's better to not break the loop upon a single ticket failure, but instead allow it to run through all tickets in case some of them do succeed?

Reply:
I initially considered not breaking the loop upon a single ticket failure, but ended up holding off because it would complicate error handling. The caller would likely want to know the errors for each ticket failure so this function would need to either aggregate the errors (perhaps using something like [go-multierror](https://github.com/hashicorp/go-multierror)) or log individual errors as they occur.

If we think that it reasonable to continue trying to redeem even after a single failure we can make that change, but my initial thought is that if one fails there is a high likelihood that the rest will fail as well because:

- If we assume that each ticket redemption is for the same sender then the recipient will be checking the same deposit and penalty escrow
- If we assume that each ticket redemption will return immediately if there is an error submitting the transaction then there is likely a deeper ETH connection issue that could cause the subsequent transactions to fail during submission as well

We could handle retries for error relating to a bad ETH connection at a layer below this so that if the recipient gets an error during a ETH related action it knows that something is definitely wrong.

Reply:
Yeah, I think this line of reasoning makes sense. My concern was indeed mainly around silly temporary errors that could be resolved with a retry, so I think having retries at a lower level could make sense. As long as ticket redemption isn't part of the critical path of video processing I think retries are a great idea.

Perhaps we should capture this thought as a SS issue?

Reply:
Tracking in #654